### ISSUE TITLE
adding init(8) like capability to initial rank 0 broker

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#518]
As discussed in [RFC 8](https://github.com/flux-framework/rfc/blob/master/spec_8.adoc), an _initial program_ is spawned on rank 0 as a stand-in for the user workload.  When the initial program terminates, the instance terminates.

In the current implementation, the default initial program is a shell.  Work can be launched interactively from the shell, or the shell can start work in a batch mode e.g. `sh -c cmd`.  When the shell terminates, the instance terminates.  The design was that this initial program concept could encompass an init(8) like program that performs user-specified initialization, such as starting a web server or loading modules, before starting the user workload.

I'd like to propose that we add a concept of a distributed run level and user-replaceable scripts associated with each run level, to run on rank 0.  The [runit](http://smarden.org/runit/index.html) init replacement has an appealingly simple design that could be used here:  It has three hardwired run levels, 1=startup, 2=muiltiuser, 3=shutdown.   Their `/bin/runit` program simply executes scripts associated with these run levels, in order, and manages run level transitions.  Runit provides process management helpers and so on to flesh out these run levels, but `/bin/runit` is not aware of them.  Our rank 0 broker could fill the role of `/bin/runit` whose logic is summarized [here](
http://smarden.org/runit/runit.8.html)

Instead of starting the initial program immediately after wireup, the broker would first run and monitor a level 1 script that could perform initialization, then transition to level 2, run the initial program as before, then if the initial program exits or shutdown is otherwise initiated, a level 3 script runs that performs shutdown.  Both the level 1 and level 3 scripts could run under a timeout.  (The level 3 timeout is already implemented as the shutdown timeout).

In he simplest case, a default script for initialization and shutdown could be provided that load/unload a default module list and otherwise behavior is as it is now.  This might be used to address issue #508.  Later, we could investigate a more full-featured run level 2 script, possibly pulling in bits of runit or some other init implementation, if appropriate.

**Comment by garlick** [comment]
Moving the hardwired list of modules out of the broker and loading them from the (user replaceable) level 1 script could solve the following issues:
* issue #508 make initial module list configurable
* issue #559 what do do on `mod_main()` error 
* issue #584 flux-broker --modopt option needs rework

This creates a new synchronization problem in some modules like the kvs, which assumes it can send requests upstream to `kvs.*` and the kvs module will already be loaded there.  This is because all the modules in the hardwired list are loaded in a given broker instance before the broker begins routing messages.  The default routing semantics are to skip over a rank if that rank doesn't have the module loaded, so something like a barrier is needed here:  kvs instances enter barrier; upon exit, start the upstream communication.

If we changed `flux module load` so that it passed
* a uuid (unique barrier name)
* a count (the number of module ranks being simultaneously loaded

to each module, that would give them enough information to execute a barrier.

The current level 1 runit script might look something like this
```sh
#!/bin/bash -e # on failure, trigger shutdown
flux module load -r all connector-local  # oops probably must be hardwired
flux module load -r all barrier
flux module load -r 0 content-sqlite
flux module load -r all kvs
flux module load -r all live
flux module load -r all mecho
flux module load -r all job
flux module load -r all wrexec
flux module load -r all resource-hwloc
```

**Comment by grondo** [comment]
Bringing this discussion over from #597 

> > Maybe runit glue can be leveraged as discussed in #518 or we could just wire up a very simple systemVinit like strategy.

> Neither of those feel like they map great to flux instance intialization because they're designed to start services and monitor them (similar idea to systemd I guess). However, I'm probably just not  wrapping my brain around the problem.


I was thinking while reading a tangent from a link on the `runit` site, that this old idea from Richard Gooch of a [need(8)](http://www.safe-mbox.com/~rgooch/linux/boot-scripts/) program would be interesting to explore for flux session initialization. There could be a way for the rank0 broker to read in the set of rc1 scripts, which notify the broker of a named "provides" or something.

A new command `flux need` could be used in scripts, and an internal broker service would track the list of named provides already loaded (like in Gooch's `init`). As described in the concept for `need(8)`, this would allow scripts to be run in any order, and allows much easier addition of scripts for 3rd party tools, as they would just `flux need` their dependencies, then load themselves, perhaps as a result of another script `flux need`-ing them.

The drawback would be more scripts, one per service, with the benefit of a much more flexible startup. I guess you could also make a faster startup by loading all scripts in parallel. If scripts were available but not run by default, you could also create minimally initialized sessions on the fly by having single-line init script like `flux need sched`.

**Comment by garlick** [comment]
Ah this is a nice, simplifying idea.  I guess having tracked the dependencies during startup, shutdown could then be orchestrated in a reverse order by `init` or broker service?

**Comment by grondo** [comment]
That would be cool, but I didn't go that far in thinking about it.

**Comment by grondo** [comment]
While we didn't address any `need(8)` like capability, is this issue resolved in principle by #597?

**Comment by garlick** [comment]
Closing.  The `need` idea is still here for history and linking from any new issues regarding rc configurability.
